src/peripherals/random.c

```diff
@@ -34,17 +42,37 @@ void rng_setup(void)
 
 u32 random_int(void)
 {
+  chMtxLock(&rng_mutex);
+
   static u32 last_value;
-  static u32 new_value;
-  u32 error_bits = 0;
-  error_bits = RNG_SR_SEIS | RNG_SR_CEIS;
-  while (new_value == last_value) {
-    /* Check for error flags and if data is ready. */
-    if (((RNG_SR & error_bits) == 0) &&
-        ((RNG_SR & RNG_SR_DRDY) == 1)) {
-      new_value = RNG_DR;
+  u32 new_value = last_value;
+  int tries = 0;
+  do {
+    u32 sr;
+    /* Check for error flags. */
+    if ((sr = RNG_SR)) {
+      if (sr & RNG_SR_SEIS) {
+        /* Seed error, we must reinitialise */
+        log_warn("random: Seed error, restarting RNG 0x%"PRIx32"\n", sr);
+        RNG_CR &= ~RNG_CR_RNGEN;
+        RNG_CR |= RNG_CR_RNGEN;
+      }
+      /* Write back as zero to clear */
+      RNG_SR = ~(sr & (RNG_SR_SEIS | RNG_SR_CEIS));
     }
+    /* Check if data is ready. */
+    if ((RNG_SR & RNG_SR_DRDY) == 0)
+      continue;
+
+    new_value = RNG_DR;
+  } while ((new_value == last_value) && (tries++ < RANDOM_MAX_TRIES));
```

Comment:
It doesn't seem to me like `RNG_SR_DRDY` is really being used here, and the timeout seems a bit fragile as its related to CPU cycles rather than anything with the peripheral?